src/descriptor/key.rs

```diff
@@ -58,6 +57,15 @@ pub enum DescriptorSecretKey {
     XPrv(DescriptorXKey<bip32::ExtendedPrivKey>),
 }
 
+/// A derived [`DescriptorPublicKey`]
+///
+/// Derived keys are guaranteed to never contain wildcards
+pub struct DerivedDescriptorKey<'secp, C: 'secp + Verification> {
```

Comment:
I think a much better name here would be `DefiniteDescriptorKey`. "Derivation" is the process of going from an extended key to a child key not from going from a key with wildcards to a key at a particular index of a wildcard.

Reply:
I am not thrilled with either names, but can't come up with a better name myself

Reply:
I think [definite vs indefinite] is the correct way of distinguishing between this property linguistically. For example in English it shows up as:

1. Definite: I'm taking **my cat** to the vet
2. Indefinite: I'm taking **a cat** to the vet (we don't know which cat)

In miniscript you have:

1. Definite: A descriptor where all derivation indexes are known
2. Indefinite: A descriptor that could be instantiated at a particular index (like "a cat" could be instantiated with any particular cat but "my cat" could not).

It might be useful to apply this naming logic to other parts of the code. There is a function named `derive` that doesn't actually derive but just sets the derivation index. Then there is a function that actually does derivation called `derive_public_key`.

[definite vs indefinite]: https://en.wikipedia.org/wiki/Definiteness

Reply:
I was confused by the current function naming you describe recently, +1 for 'definite'.

Reply:
I think renaming `derive` to `set_derivation_index` would probably be better than `make_definite` or something. But I do think definite and indefinite in the type names make sense.

Reply:
Maybe `Concrete`/`Abstract`?

Reply:
I'll let you guys discuss this, let me know when you are settled with one and I'll update my PR.

I personally don't have any preference.